src/py/flwr/server/superlink/state/in_memory_state.py

```diff
@@ -210,17 +215,32 @@ def get_nodes(self, run_id: int) -> Set[int]:
         If the provided `run_id` does not exist or has no matching nodes,
         an empty `Set` MUST be returned.
         """
-        if run_id not in self.run_ids:
-            return set()
-        return self.node_ids
+        with self.lock:
+            if run_id not in self.run_ids:
+                return set()
+            current_time = time.time()
+            return {
+                node_id
+                for node_id, online_until in self.node_ids.items()
+                if online_until > current_time
+            }
 
     def create_run(self) -> int:
         """Create one run."""
         # Sample a random int64 as run_id
-        run_id: int = int.from_bytes(os.urandom(8), "little", signed=True)
+        with self.lock:
+            run_id: int = int.from_bytes(os.urandom(8), "little", signed=True)
 
-        if run_id not in self.run_ids:
-            self.run_ids.add(run_id)
-            return run_id
+            if run_id not in self.run_ids:
+                self.run_ids.add(run_id)
+                return run_id
         log(ERROR, "Unexpected run creation failure.")
         return 0
+
+    def acknowledge_ping(self, node_id: int, ping_interval: float) -> bool:
+        """Acknowledge a ping received from a node, serving as a heartbeat."""
+        with self.lock:
+            if node_id in self.node_ids:
+                self.node_ids[node_id] = time.time() + ping_interval
```

Comment:
why add the `ping_interval` to the timestamp here? wouldn't it be better to use the interval just when doing `get_nodes()